### PR TITLE
Implement HTTP typed resource responses

### LINKED ISSUE #478: [Service Based Implementation] Implement HTTP typed resource responses
**Description:**
Similar to HTTP library we need to introduce records based on status codes, This also should support the payload annotation to return the body directly. 

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Thevakumar-Luheerathan**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: ballerina-tests/main.bal [hunk 1]
```diff
@@ -16,6 +16,31 @@ service "hello" on ep {
...
+             headers: {
+                 "content-type": "application/json+id",
+                 "Location": "/myServer/084230"
+             }};
+         return err;
+     }
+     resource function post httpResTest4(@af:Payload string greeting) returns @af:HTTPOutput af:InternalServerError {
+              af:InternalServerError err = {};
+              return err;
+          }
```
Comment (anuruddhal) [inline ballerina-tests/main.bal:43]: Shall we fix formatting issues?

File: ballerina/http_status_code_types.bal [hunk 2]
```diff
@@ -0,0 +1,932 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anuruddhal) [inline ballerina/http_status_code_types.bal:1]: Fix license header hear

File: native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java [hunk 3]
```diff
@@ -81,6 +82,50 @@ public void notifySuccess(Object result) {
 //        }
         if (this.annotations.get(0).equals("QueueOutput") || this.annotations.get(0).equals("CosmosDBOutput")) {
             mapValue.put(StringUtils.fromString("outMsg"), result);
+        } else if (this.annotations.get(0).equals("HTTPOutput")) {
+            if (isHTTPResponse(result)) {
+                BObject status = (BObject) (((BMap) result).get(StringUtils.fromString("status")));
```
Comment (xlight05) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:87]: Shall we extract the `(BMap) result` into a variable?

File: native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java [hunk 4]
```diff
@@ -81,6 +82,50 @@ public void notifySuccess(Object result) {
 //        }
         if (this.annotations.get(0).equals("QueueOutput") || this.annotations.get(0).equals("CosmosDBOutput")) {
             mapValue.put(StringUtils.fromString("outMsg"), result);
+        } else if (this.annotations.get(0).equals("HTTPOutput")) {
```
Comment (anuruddhal) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:85]: Make constant as the first argument
```java
"HTTPOutput".equals(this.annotations.get(0))
```

**Review by anuruddhal — CHANGES_REQUESTED** [review]

File: native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java [hunk 5]
```diff
@@ -81,6 +82,50 @@ public void notifySuccess(Object result) {
...
+                BMap<BString, Object> bodyMap =
+                        ValueCreator.createMapValue(TypeCreator.createMapType(PredefinedTypes.TYPE_ANYDATA));
+                bodyMap.put(StringUtils.fromString("statusCode"), statusCode);
+
+                if (((BMap) result).containsKey(StringUtils.fromString("body"))) {
+                    Object body = ((BMap) result).get(StringUtils.fromString("body"));
+                    bodyMap.put(StringUtils.fromString("body"), body);
+                }
+
+                if (((BMap) result).containsKey(StringUtils.fromString("headers"))) {
```
Comment (xlight05) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:100]: shall we extract this(and other) complex logic into a function to improve the readability?

File: native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java [hunk 6]
```diff
@@ -81,6 +82,50 @@ public void notifySuccess(Object result) {
...
+                    Object headers = ((BMap) result).get(StringUtils.fromString("headers"));
+                    if (!((BMap) headers).containsKey(StringUtils.fromString("content-type"))) {
+                        ((BMap) headers).put(StringUtils.fromString("content-type"),
```
Comment (xlight05) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:103]: shall we extract the converted headers bmap into a variable?

Comment (xlight05) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:103]: shouldn't we do the check after converting content-type to lowercase? (to account for caps issues)

File: ballerina-tests/tests/test.bal [hunk 7]
```diff
@@ -12,6 +12,52 @@ function testBaseDot() returns error? {
    test:assertEquals(resp, expectedResp);
 }
 
+@test:Config { }
+function httpResTest1() returns error? {
+   final http:Client clientEndpoint = check new ("http://localhost:3000");
+   string jsonFilePath = "./tests/resources/httpResTest1.json";
+   json readJson = check io:fileReadJson(jsonFilePath);
+   json resp = check clientEndpoint->post("/post-hello-httpResTest1", readJson);
```
Comment (xlight05) [inline ballerina-tests/tests/test.bal:20]: do we need to introduce multiple jsons? can't we reuse existing json we have, and reuse it in the test cases?

**Review by xlight05 — CHANGES_REQUESTED** [review]

File: native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java [hunk 8]
```diff
@@ -42,10 +44,23 @@
  */
 public class FunctionCallback implements Callback {
 
+    private static final String QUEUE_OUTPUT = "QueueOutput";
```
Comment (xlight05) [inline native/src/main/java/io/ballerina/stdlib/azure/functions/FunctionCallback.java:47]: Shall we move these into Constants Class?

**Review by xlight05 — APPROVED** [review]

**Pull request merged**